// @fesk/live-server-plugin/index.js
const waitOn = require('wait-on');
const fs = require('fs');

const PORT = process.env.LIVE_SERVER_PORT || 5000;

const params = {
  port: PORT,
  root: 'dist',
  open: false,
  wait: 100,
  logLevel: 0,
  middleware: [
    function(req, res, next) {
      if (
        req.url.endsWith('.css') === false &&
        req.url.endsWith('.js') === false &&
        fs.existsSync(`${process.cwd()}/dist/assets/manifest.json`) === false
      ) {
        console.log('Waiting for manifest...');
        waitOn(
          {
            resources: [`${process.cwd()}/dist/assets/manifest.json`],
          },
          function() {
            res.setHeader(
              'Cache-Control',
              'private, no-cache, no-store, must-revalidate'
            );
            res.setHeader('Expires', '-1');
            res.setHeader('Pragma', 'no-cache');
            next();
          }
        );
      } else {
        res.setHeader(
          'Cache-Control',
          'private, no-cache, no-store, must-revalidate'
        );
        res.setHeader('Expires', '-1');
        res.setHeader('Pragma', 'no-cache');
        next();
      }
    },
  ],
};

module.exports = params;
