// metalsmith-atomic-design/lib/atomic-twig-templates.js
const pluginKit = require('metalsmith-plugin-kit');
const twigTransform = require('metalsmith-twig-transform');

/**
 * Atomic twig templates.
 *
 * This will take atoms, molecules and other components or partials that do not have a layout and
 * apply a pattern library layout to them. This should be applied to a copied file tree, otherwise it
 * may break partial includes, documentation etc.
 *
 * @param opts
 * @returns {atomicTwigTemplatesPlugin}
 */
function atomicTwigTemplates(opts) {
  const config = Object.assign(
    {
      layouts: 'lib/layouts',
      route: 'pattern-library',
      defaultLayout: 'pattern-library.twig',
      structure: ['atoms', 'molecules', 'organisms', 'pages'],
      reporter: null,
    },
    opts
  );

  function atomicTwigTemplatesPlugin(files, metalsmith, done) {
    return twigTransform({
      pattern:
        config.route + '/+(' + config.structure.join('|') + ')/**/*.twig',
      layouts: config.layouts,
      defaultLayout: config.defaultLayout,
      reporter: config.reporter,
    })(files, metalsmith, done);
  }

  pluginKit.renameFunction(atomicTwigTemplatesPlugin, 'atomic-twig-templates');

  return atomicTwigTemplatesPlugin;
}

module.exports = atomicTwigTemplates;
